src/DotNet.Testcontainers/Containers/ResourceReaper.cs

```diff
@@ -0,0 +1,288 @@
+namespace DotNet.Testcontainers.Containers
+{
+  using System;
+  using System.IO;
+  using System.Net;
+  using System.Net.Sockets;
+  using System.Text;
+  using System.Threading;
+  using System.Threading.Tasks;
+  using DotNet.Testcontainers.Builders;
+  using DotNet.Testcontainers.Clients;
+  using DotNet.Testcontainers.Configurations;
+
+  public sealed class ResourceReaper : IAsyncDisposable
+  {
+    public const string ResourceReaperSessionLabel = TestcontainersClient.TestcontainersLabel + ".resource-reaper-session";
+
+    private static readonly SemaphoreSlim DefaultLock = new SemaphoreSlim(1, 1);
+
+    private static ResourceReaper defaultInstance;
+
+    private readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+    private readonly TestcontainersContainer resourceReaperContainer;
+
+    private Task connectionTask;
+
+    private bool disposed;
+
+    private ResourceReaper(Guid sessionId)
+    {
+      this.resourceReaperContainer = new TestcontainersBuilder<TestcontainersContainer>()
+        .WithName($"testcontainers-ryuk-{sessionId:D}")
+        .WithImage(RyukImage)
+        .WithAutoRemove(true)
+        .WithCleanUp(false)
+        .WithExposedPort(RyukPort)
+        .WithPortBinding(0, RyukPort)
+        .WithBindMount("/var/run/docker.sock", "/var/run/docker.sock", AccessMode.ReadOnly)
+        .Build();
+
+      this.SessionId = sessionId;
+    }
+
+    /// <summary>
+    /// Gets the default <see cref="ResourceReaper" /> session id.
+    /// </summary>
+    /// <remarks>
+    /// The default <see cref="ResourceReaper" /> will start either on <see cref="GetAndStartDefaultAsync(System.Threading.CancellationToken)" />
+    /// or if a <see cref="ITestcontainersContainer" /> is configured with <see cref="ITestcontainersBuilder{TDockerContainer}.WithCleanUp" />.
+    /// </remarks>
+    public static Guid DefaultSessionId { get; }
+      = Guid.NewGuid();
+
+    /// <summary>
+    /// Gets the <see cref="ResourceReaper" /> session id.
+    /// </summary>
+    public Guid SessionId { get; }
+
+    /// <summary>
+    /// Sets image for Ryuk container.
+    /// </summary>
+    public static string RyukImage { get; set; } = "ghcr.io/psanetra/ryuk:2021.12.20";
```

Comment:
The `RyukPort` property was nice to have to write one of the tests. Otherwise it is probably not really useful. I think especially the `RyukImage` property is needed as I expect some users need to point to some internal container registry if they are not allowed to use arbitrary registries. Alternatively we could provide a special environment variable to set a custom image name.

I could rewrite the test to use a different image instead of a port. Then we could get rid of the port property.

The `InitTimeout` is also mostly nice to have for tests. I don't know if someone maybe wants to change it in a real environment. (Maybe downloading the image might be too slow in a cicd pipeline...)

Reply:
Then lets overload. What about something like `GetAndStartNewAsync(Guid sessionId, string ryukImage = RyukImage, string ryukPort = RyukPort, CancellationToken ct = default)`?

Reply:
ok, should we remove the `InitTimeout` completely and expect the user to provide such a timeout if they want to have one? (I think it should be the default to have one. At least for the Default-ResourceReaper.)

Reply:
No, we can keep the timeout if you like. Just add it as an additional parameter (maybe `initTimeoutInMillisecond`), or what do you think? Any better idea? I just think the static configuration is odd.

Reply:
Done